### ISSUE TITLE
Discovery queue customization

### ISSUE THREAD

**Issue by antonkurkin** [ValveSoftware/steam-for-linux#4500]
#### Your system information
- Steam client version: 1465929127
- Distribution: Debian Sid
- Opted into Steam client beta?: Yes
- Have you checked for system updates?: Yes
#### Please describe your issue in as much detail as possible:

There is new "Customize your queue" button on discovery queue page http://store.steampowered.com/explore/ and in customization menu there is no OS filtering. After some changes in there my queue is now full of windows-only titles and I can't get back to what it was.
I'm not sure how I did filtering for linux support in the first place. I believe there was another customization menu some time before that

**Comment by ragsu** [comment]
I am having the same kind of issue. I can't filter out the non-linux games at all anymore. Been that way for about over a week. There was already some others discussing about it at the forums:
http://steamcommunity.com/app/221410/discussions/0/358415206081937798/

**Comment by GhettoGirl** [comment]
This "new" bug is very annoying. Since the queue customization feature, I'm only seeing windows-only games in my queue ;( The discover queue is now literally garbage.
What did you guys do that the OS filter broke so hard? Please fix ASAP, thank you.

**Comment by soverton** [comment]
Same problem here. Is there any sort of work-around? Discovery queue is useless to me as it stands now.

**Comment by antonkurkin** [comment]
And it looks like "Popular New Releases"/"Top Sellers"/"Upcoming"/"Specials" default filter is now set to Windows+Mac. Yep, linux/steamOS isn't even selected in "See more", not that it really matters of course

**Comment by BlaXpirit** [comment]
Some people say the queue still works fine for them (they see Linux games). It is possible that the queue shows only Windows games to people who had previously entirely depleted the queue.

Not much else to add, but I will note that this is a big problem for me.

**Comment by isage** [comment]
It starts showing Windows-only games after customization. Resetting filters
afterwards does not fix it.

**Comment by tesfabpel** [comment]
I think it is a store bug, since it affects me on the browser too...
Seems like when you customize your Discovery Queue it will clear the Linux / SteamOS filter that likely was there before...

**Comment by alexmaras** [comment]
Yep, I had a queue with only SteamOS/Linux stuff, customised it to remove a couple of tags, it need to reload the queue and since then there's been no restrictions on platform. I did all this on the browser, so it must be an issue with the new queue filters on the store.

**Comment by benopotamus** [comment]
I have this problem too. It does still showing Linux games, but it isn't showing _only_ Linux games anymore.

**Comment by 1338h4x** [comment]
The queue _used_ to only me Linux games. This is a pretty bad regression.

**Comment by eNTi** [comment]
please fix this. the queue is basically useless now.

**Comment by fuxoft** [comment]
Seconding this. I have ONLY Linux and 95% of my queues are now non-Linux games, i.e. the queues are useless to me. Before, I was shown only Linux games. Now I have no idea where to even configure that I have Linux...?

**Comment by tesfabpel** [comment]
I've tried going here: https://store.steampowered.com/account/preferences/
and I found there is a section called something like "Platform settings" where you can select the platforms you're interested in... (I didn't have any platform selected...)
Anyway nothing changed yet... I suppose I have to wait a little bit (if it works :P)...

**Comment by soverton** [comment]
Has any action been taken on this, or has it been filed under "Who cares"?

**Comment by fuxoft** [comment]
tesfabfel, I tried that, selected Linux only and now, 12 hours later, I am still seeing Windows games everywhere... :(

**Comment by antonkurkin** [comment]
I found "Browse All New Releases" button today on the store's main page. There is "Your New Release Queue" and it is actually linux-only (at least for me). Queue on the store's main page still doesn't filter by OS though.

**Comment by philipl** [comment]
And as soon as you customize it, you're back to seeing windows/mac games.

**Comment by fuxoft** [comment]
Is this even acknowledged as a bug? Or are we doing something wrong?

**Comment by antonkurkin** [comment]
Actually, I believe it's more of feature request (or a bug if you like it more) for steam store (web page) than real bug of steam client (application) for linux. I only thought about it after I posted the issue. But this report was tagged as "reviewed" so it should have passed some perfunctory check.

**Comment by BlaXpirit** [comment]
I hate this so much.

Every time someone says *"Oh cool, such and such game is now on Linux!"*, all I can think of is *"Damn you Valve, I used to have a reliable way to find out about these"*

Not to mention having to wade through the "early access" and even **completely unaccessible** games in the "new releases" section.

**Comment by at46** [comment]
@BlaXpirit You can check https://steamdb.info/linux/ for new games with Linux support.

**Comment by BlaXpirit** [comment]
@at46 Yes, except someone has to try out the games and add them to the list first (as [I often do](https://github.com/SteamDatabase/SteamLinux/commits?author=BlaXpirit)). And there still are ["Game possibly works: 621"] (https://steamdb.info/linux/hints/).

Also, some partially working games never end up there.

Not reliable, unfortunately.

**Comment by aaronfranke** [comment]
I have the same issue, the OS filter settings do not work, and the Discovery queue is affected as well. I can see non-Linux games even after activating the setting, as seen here: https://imgur.com/a/2f1iQ

**Comment by ememoho** [comment]
TL;DR - An illustrated guide to re-enable OS filters for your account: https://imgur.com/a/kYYbd

Since this issue has, apparently, only hit those curious enough to try out the new filtering system, and did not affect other users (anecdotal evidence), I figured the OS-filtering capabilities were still intact on the backend. So I've poked around a bit with the Network Console in the Firefox's WebDev tools, and here's what I discovered.

Clicking the "Customize your queue" button sends a
`GET http://store.steampowered.com/explore/discoveryqueuesettings`

In response to this, Steam sends a JSON wherein alongside the html code for the settings popup window and popular tags, you can find the "Settings" object with the current DQ options:
```javascript
"settings": {
    "os_win": null,
    "os_mac": null,
    "os_linux": null,
    "full_controller_support": null,
    "native_steam_controller": null,
    "include_coming_soon": 0,
    "excluded_tagids": [],
    "exclude_early_access": 1,
    "exclude_videos": null,
    "exclude_software": 1
  }
```

Note how the os_ settings are all null-ed out.
This is because when you click "OK" in the customization window, a
`POST http://store.steampowered.com/explore/updatediscoveryqueuesettings`
request goes out with the following form data:

```
sessionid: xxxxxxxxxxxxxxxxxxxxxxxx
settings: {"exclude_early_access":true,"exclude_software":true,"include_coming_soon":true,"excluded_tagids":[]}"
queuetype: 0
```
A much smaller set of parameters is sent when you update the settings.

Next thing I've tried is the "Edit and Resend" button in the Network Console, and my resulting request looked like this:
`POST http://store.steampowered.com/explore/updatediscoveryqueuesettings`
Request body:
`sessionid=xxxxxxxxxxxxxxxxxxxxxxxx&settings=%7b%22exclude_early_access%22%3atrue%2c%22exclude_software%22%3atrue%2c%22include_coming_soon%22%3atrue%2c%22excluded_tagids%22%3a%5b%5d%2c%22os_win%22%3a0%2c%22os_mac%22%3a0%2c%22os_linux%22%3a1%7d&queuetype=0`
which is the same set with the addition of my preferred OS settings (win:0, mac:0, linux:1).

And Voila! The settings are back.
Refresh the Discovery Queue page, and enjoy the *relevant* results.

This is definitely something that should be fixed, and fixed on the Store's frontend (not specific to steam-for-linux), but in case anyone is looking for a workaround - look no further.

**Comment by GhettoGirl** [comment]
@ememoho The POST request worked for me, and I finally see Linux games again in my library. Thanks for this awesome finding :heart: 

**Valve, PLEASE FIX ASAP!** Thanks

**Comment by Suppen** [comment]
Beautiful! Worked for me too! Should be very easy to fix for Valve. When I contacted Steam Support about it a few months ago, they said "We're looking into it"

**Comment by BlaXpirit** [comment]
Here is an easier way to send that POST request, for a workaround.

Go to any page on the Steam Store, open the browser's JS console (usually F12), paste and customize this code, press Enter:

```javascript
jQuery.post('//store.steampowered.com/explore/updatediscoveryqueuesettings', {
    sessionid: g_sessionID, queuetype: 0,
    settings: JSON.stringify({
        exclude_early_access: true, exclude_software: true,
        exclude_videos: true, include_coming_soon: false,
        excluded_tagids: [],
        os_win: 0, os_mac: 0, os_linux: 1
    })
})
```

**Comment by tesfabpel** [comment]
@Plagman @ememoho has found the solution... 
The fix should be very simple to implement on the frontend.

**Comment by fuxoft** [comment]
@BlaXpirit Clarification: It does not work on ANY page in the store. E.g. the game profile pages don't contain jQuery.

**Comment by ragsu** [comment]
I get "bad request" when doing this. I tried many times already and it does not change. For some reason my edited POST 0.04KB next to the original 0.89KB. What am I doing wrong? Thank you @ememoho for your great work!

**Comment by Suppen** [comment]
I also got "400 Bad request" and "401 Not authorized" a few times. Figured out I was URL-encoding too much. Make sure you don't URL-encode the ='s and &'s

**Comment by ragsu** [comment]
Apparently I had removed the queuetype accidentally so that was fixed and I got the change through. However even if the options changed, I see no difference in the discovery queue nor the store. @BlaXpirit where do I paste this code in the console? I want to try out this way too.

**Comment by RomanHargrave** [comment]
It seams that the `os_` queue settings have silently been "fixed" as it were, and no longer affect the queue.

Back to receiving a windfall of Windows-only unity asset flips and boring overhyped games.

**Comment by Suppen** [comment]
@RomanHargrave Looks like you are right. I got windows-only bullcrap in my queue again...

**Comment by Suppen** [comment]
I still get {os_win: 0, os_mac: 0, os_linux: 1} on the response to the "customize" HTTP request, so now it is a backend problem too

**Comment by Suppen** [comment]
{os_win: 0, os_mac: 0, os_linux: 1} appears to no longer be processed on the server if I modify the POST request as described earlier in this thread.

After changing my customization, it now says {os_win: null, os_mac: null, os_linux: null}, even after going through the above procedure

**Comment by aaronfranke** [comment]
At this point Valve simply doesn't care anymore about Linux gaming. This is completely pathetic. Not only this, but also the lack of Vive/SteamVR support, the lack of Alien Swarm and Left4Dead 1 on Linux, the abundant delay or cancellation of Valve-sponsored Linux ports of games, and the complete lack of any media center features such as Netflix on SteamOS.

**Comment by ememoho** [comment]
@RomanHargrave 
That is odd. My Discovery queue has been Linux-friendly since POSTing the OS filters.
Just now I've clicked through 12 Linux-compatible items, generated another one, and it's all good again.
Is it the same within the Steam client and a generic web browser?

Are you sure you have not exhausted the whole SteamOS/Linux library? In one of the recent seasonal sales, Steam was giving away trading cards for completing discovery queues (limited to 3 per day), and it automatically threw 3 cards at me, because there simply was not enough non-viewed offers in the catalogue (with my OS preferences applied) to populate a queue. They could have possibly tweaked that bit.

@Suppen 
Note: If you save the customization settings via the webpage itself, you reset the "hidden" filters.

**Comment by pwcjr** [comment]
I abandoned Steam discovery queues a bit ago after noticing them full of non-Linux games, but after finding this issue, I gave the POST described above a shot. No dice, unfortunately. I still see almost no Linux-compatible games via browser or Steam Linux client. *le sigh*

**Comment by PVince81** [comment]
Strange thing. For me the discovery queue was always Linux only for more than a year. It's only since a few days, possibly with the start of the Halloween sale that it started to show only Windows titles. I thought that maybe they added a special override because the queue only showed Halloween sale title. But now that the sale is over it still only stuck on non-Linux titles.

This is sad because the queue used to be quite useful for discovering new titles.

This is on build from Oct 13th 2016 at 00:48:25.
Running on openSUSE Tumbleweed.

**Comment by benopotamus** [comment]
The POST method worked for me (I did it early October). I just checked and everything on the store front page is Linux(SteamOS) compatiable.

**Comment by RomanHargrave** [comment]
Odd. I wouldn't be surprise at this point if three different variations of the store backend are going around right now.

**Comment by ragsu** [comment]
After I did the POST fix it took a few days and the steam store was fixed. Now all I see is Linux games.

**Comment by stratus-ss** [comment]
I usually only use the steam client... this should be configurable in the client itself

I did the post, I got the response and I see os_linux:1 but I am still getting windows only games... guess it didnt work for me

**Comment by kisak-valve** [comment]
Hello, is anyone still seeing an issue with the discovery queue and the desired platforms set in your [account's preferences](https://store.steampowered.com/account/preferences/)?

**Comment by at46** [comment]
Everything is fine for me with the discovery queue.

**Comment by oprypin** [comment]
The behavior of the discovery queue definitely went back to its prior state at some point, and this particular issue is gone.

**Comment by kisak-valve** [comment]
Thanks for the feedback, closing as fixed some time ago.
